pkg/limits/ingest_limits.go

```diff
@@ -389,62 +353,57 @@ func (s *IngestLimits) GetAssignedPartitions(_ context.Context, _ *logproto.GetA
 	return &resp, nil
 }
 
-// GetStreamUsage implements the logproto.IngestLimitsServer interface.
+// ExceedsLimits implements the logproto.IngestLimitsServer interface.
 // It returns the number of active streams for a tenant and the status of requested streams.
-func (s *IngestLimits) GetStreamUsage(_ context.Context, req *logproto.GetStreamUsageRequest) (*logproto.GetStreamUsageResponse, error) {
-	// Get the cutoff time for active streams
-	cutoff := time.Now().Add(-s.cfg.WindowSize).UnixNano()
-
-	// Calculate the rate window cutoff in nanoseconds
-	rateWindowCutoff := time.Now().Add(-s.cfg.RateWindow).UnixNano()
-
-	// Count total active streams for the tenant
-	// across all assigned partitions and record
-	// the streams that have been seen within the
-	// window
+func (s *IngestLimits) ExceedsLimits(_ context.Context, req *logproto.ExceedsLimitsRequest) (*logproto.ExceedsLimitsResponse, error) {
 	var (
-		activeStreams uint64
-		totalSize     uint64
+		lastSeenAt = time.Now()
+		// Use the provided lastSeenAt timestamp as the last seen time
+		recordTime = lastSeenAt.UnixNano()
+		// Calculate the cutoff for the window size
+		cutoff = lastSeenAt.Add(-s.cfg.WindowSize).UnixNano()
+		// Get the bucket for this timestamp using the configured interval duration
+		bucketStart = lastSeenAt.Truncate(s.cfg.BucketDuration).UnixNano()
+		// Calculate the rate window cutoff for cleaning up old buckets
+		bucketCutoff = lastSeenAt.Add(-s.cfg.RateWindow).UnixNano()
+		// Calculate the max active streams per tenant per partition
+		maxActiveStreams = uint64(s.limits.MaxGlobalStreamsPerUser(req.Tenant) / s.cfg.NumPartitions)
+		// Create a map of streams per partition
+		streams = make(map[int32][]Stream)
 	)
 
-	// If the stream is written into a partition we are
-	// assigned to and has been seen within the window,
-	// it is an active stream.
-	unknownStreams := req.StreamHashes
-
-	s.metadata.Usage(req.Tenant, func(partitionID int32, stream Stream) {
-		if assigned := s.partitionManager.Has(partitionID); !assigned {
-			return
-		}
+	for _, stream := range req.Streams {
+		partitionID := int32(stream.StreamHash % uint64(s.cfg.NumPartitions))
```

Comment:
I think we are missing a check if the `partitionID` is also assigned?

Reply:
Yes indeed. Do you think we can silently skip the streams here then? Or should er report this as an error to the frontend to retry on the right partition consumer?

Reply:
I think for now we need to drop the streams (i.e. reject them). If we skip, then the streams are accepted, and this will allow limits to be exceeded.

Reply:
Sorry I have to precise a bit my question here. We already drop these streams, however we just write a warn log message in addition. Do we want to bubble up an error to the frontend to retry on other partition consumers?

Reply:
Ah OK, to be clear, the streams **also** need to be rejected (not just dropped).